cores/esp32/HWCDC.cpp

```diff
@@ -260,35 +376,63 @@ size_t HWCDC::write(const uint8_t *buffer, size_t size)
-        size = 0;
+    if(!isCDC_Connected()) {
+        // just pop/push RingBuffer and apply FIFO policy
+        flushTXBuffer(buffer, size);
     } else {
-        to_send -= space;
-        so_far += space;
-        // Now trigger the ISR to read data from the ring buffer.
-        usb_serial_jtag_ll_ena_intr_mask(USB_SERIAL_JTAG_INTR_SERIAL_IN_EMPTY);
+        size_t space = xRingbufferGetCurFreeSize(tx_ring_buf);
+        size_t to_send = size, so_far = 0;
 
-        while(to_send){
-            if(max_size > to_send){
-                max_size = to_send;
-            }
-            // Blocking method, Sending data to ringbuffer, and handle the data in ISR.
-            if(xRingbufferSend(tx_ring_buf, (void*) (buffer+so_far), max_size, tx_timeout_ms / portTICK_PERIOD_MS) != pdTRUE){
-                size = so_far;
-                break;
-            }
-            so_far += max_size;
-            to_send -= max_size;
+        if(space > size){
+            space = size;
+        }
+        // Non-Blocking method, Sending data to ringbuffer, and handle the data in ISR.
+        if(space > 0 && xRingbufferSend(tx_ring_buf, (void*) (buffer), space, 0) != pdTRUE){
+            size = 0;
+        } else {
+            to_send -= space;
+            so_far += space;
             // Now trigger the ISR to read data from the ring buffer.
-            usb_serial_jtag_ll_ena_intr_mask(USB_SERIAL_JTAG_INTR_SERIAL_IN_EMPTY);
+            usb_serial_jtag_ll_txfifo_flush();
+            if(connected) usb_serial_jtag_ll_ena_intr_mask(USB_SERIAL_JTAG_INTR_SERIAL_IN_EMPTY);
+            // tracks CDC trasmission progress to avoid hanging if CDC is unplugged while still sending data
+            size_t last_toSend = to_send; 
+            uint32_t tries = tx_timeout_ms; // waits 1ms per sending data attempt, in case CDC is unplugged
+            while(connected && to_send){
+                space = xRingbufferGetCurFreeSize(tx_ring_buf);
+                if(space > to_send){
+                    space = to_send;
+                }
+                // Blocking method, Sending data to ringbuffer, and handle the data in ISR.
+                if(xRingbufferSend(tx_ring_buf, (void*) (buffer+so_far), space, tx_timeout_ms / portTICK_PERIOD_MS) != pdTRUE){
+                    size = so_far;
+                    log_w("write failed due to ring buffer full - timeout");
+                    break;
+                }
+                so_far += space;
+                to_send -= space;
+                // Now trigger the ISR to read data from the ring buffer.
+                usb_serial_jtag_ll_txfifo_flush();
+                if(connected) usb_serial_jtag_ll_ena_intr_mask(USB_SERIAL_JTAG_INTR_SERIAL_IN_EMPTY);
+                if(last_toSend == to_send) {
+                    // no progress in sending data... USB CDC is probably unplugged
+                    tries--;
+                    delay(1);
```

Comment:
same comment about delay, unless you are waiting on ISR, in which case it's OK

Reply:
Yes, this is just to wait for the ISR (HOST to request all data). 
It is related to the Host being busy and not allowing the Device to send a very, very long sized char *buffer. I Can set time out to 30 seconds for instance and it will wait the host to process the data and ask for more data not breaking Write().